### ISSUE TITLE
Alternative semantics of For

### ISSUE THREAD

**Issue by crusso** [caffeinelabs/motoko#852]
Currently:

```
The `for`-expression is syntactic sugar for

```bnf
for ( <pat> in <exp1> ) <exp2> :=
  {
    let x = <exp1>;
    label l loop {
      switch (x.next()) {
        case (? <pat>) <exp2>;
/*    case (? _) continue l;  // a */
/*    case (? _) break l;   // b */
        case (null) break l;
      }
    }
  }

where `x` is a fresh identifier.
```

* If we add the line //a  comments, it would filter instead of trapping (like Swift's for case let (apparently))
* If we add line //b in comments, it would exit instead of trapping 
(like Rusts `while let` https://doc.rust-lang.org/rust-by-example/flow_control/while_let.html)

Both are sensible, and perhaps better than trapping.

We could also consider something adding `while (pat in exp) exp` if that fits the grammar.

**Comment by crusso** [comment]
FYI @ggreif (since you were consider `for (case <pat> in <exp>) { }`)

**Comment by ggreif** [comment]
Yeah, my suggestion is in https://github.com/dfinity/motoko/issues/3690.

There were also voices mentioning `if (case <pat> = <exp>) { ... } else ...` and `while (case ...`.
